File: codes/javascript/src/string_interpolation.js
const jsType = { "Null": "Null", "Boolean": "Boolean", "String": "String", "Numeric": "Numeric", "Object": "Object", "Array": "Array", "Function": "Function" };

const isNull = (anything) => (((Object.prototype.toString.call(anything) === "[object Null]") && (anything === null)) || ((Object.prototype.toString.call(anything) === "[object Undefined]") && (anything === undefined)));

const isBoolean = (anything) => ((Object.prototype.toString.call(anything) === "[object Boolean]") && ((anything === true) || (anything === false)));

const isString = (anything) => (Object.prototype.toString.call(anything) === "[object String]");

const isNumeric = (anything) => ((Object.prototype.toString.call(anything) === "[object Number]") && (Number.isNaN(anything) === false) && (Number.isFinite(anything) === true));

const isObject = (anything) => (Object.prototype.toString.call(anything) === "[object Object]");

const isArray = (anything) => ((Object.prototype.toString.call(anything) === "[object Array]") && (Array.isArray(anything) === true));

const isFunction = (anything) => (Object.prototype.toString.call(anything) === "[object Function]");

const getType = (anything) => ((isNull(anything) === true) ? jsType.Null : ((isBoolean(anything) === true) ? jsType.Boolean : ((isString(anything) === true) ? jsType.String : ((isNumeric(anything) === true) ? jsType.Numeric : ((isObject(anything) === true) ? jsType.Object : ((isArray(anything) === true) ? jsType.Array : ((isFunction(anything) === true) ? jsType.Function : Object.prototype.toString.call(anything))))))));

const jsonStringify = (anything, { pretty = false } = {}) => {
    // custom JSON.stringify() function jsonStringifyV3
    const indent = " ".repeat(4);
    let indentLevel = 0;
    const jsonStringifyInner = (anythingInner) => {
        const anythingInnerType = getType(anythingInner);
        if (anythingInnerType === jsType.Null) return "null";
        if (anythingInnerType === jsType.String) return `"${anythingInner}"`;
        if ((anythingInnerType === jsType.Numeric) || (anythingInnerType === jsType.Boolean)) return `${anythingInner}`;
        if (anythingInnerType === jsType.Object) {
            if (Object.keys(anythingInner).length === 0) return "{}";
            indentLevel += 1;
            let result = ((pretty === true) ? (`{\n${indent.repeat(indentLevel)}`) : "{ ");
            Object.entries(anythingInner).forEach(([objectKey, objectValue], objectEntryIndex) => {
                result = `${result}"${objectKey}": ${jsonStringifyInner(objectValue)}`;
                if ((objectEntryIndex + 1) !== Object.keys(anythingInner).length) {
                    result = `${result}${((pretty === true) ? (`,\n${indent.repeat(indentLevel)}`) : ", ")}`;
                }
            });
            indentLevel -= 1;
            result = `${result}${((pretty === true) ? (`\n${indent.repeat(indentLevel)}}`) : " }")}`;
            return result;
        }
        if (anythingInnerType === jsType.Array) {
            if (anythingInner.length === 0) return "[]";
            indentLevel += 1;
            let result = ((pretty === true) ? (`[\n${indent.repeat(indentLevel)}`) : "[");
            anythingInner.forEach((arrayItem, arrayItemIndex) => {
                result = `${result}${jsonStringifyInner(arrayItem)}`;
                if ((arrayItemIndex + 1) !== anythingInner.length) {
                    result = `${result}${((pretty === true) ? (`,\n${indent.repeat(indentLevel)}`) : ", ")}`;
                }
            });
            indentLevel -= 1;
            result = `${result}${((pretty === true) ? (`\n${indent.repeat(indentLevel)}]`) : "]")}`;
            return result;
        }
        if (anythingInnerType === jsType.Function) return "[object Function]";
        return anythingInnerType;
    };
    return jsonStringifyInner(anything);
};

console.log("// Template literals / Template strings (String Interpolation) in JavaScript");

const myName = "Alisa";
const myAge = 25;
console.log(`my name is ${myName} and I am ${myAge}.`);

console.log(`1 + 2 + 3 + 4 = ${1 + 2 + 3 + 4}`);

function getRectangleArea(rectangleWidth, rectangleLength) {
    return (rectangleWidth * rectangleLength);
}
console.log(`getRectangleArea(7, 5): ${getRectangleArea(7, 5)}`);

const anyString = "foo";
console.log(`any string: ${jsonStringify(anyString)}`);

const anyNumeric = 123;
console.log(`any numeric: ${jsonStringify(anyNumeric)}`);

const anyBoolean = true;
console.log(`any boolean: ${jsonStringify(anyBoolean)}`);

const anyNull = null;
console.log(`any null: ${jsonStringify(anyNull)}`);

const anyArray = [1, 2, 3];
console.log(`any array: ${jsonStringify(anyArray)}`);
console.log(`any array first element: ${jsonStringify(anyArray?.[0])}`);

const anyObject = { "foo": "bar" };
console.log(`any object: ${jsonStringify(anyObject)}`);
console.log(`any object foo value: ${jsonStringify(anyObject?.foo)}`);

File: codes/lua/string_interpolation.lua
js_like_type = { ["Null"] = "Null", ["Boolean"] = "Boolean", ["String"] = "String", ["Numeric"] = "Numeric", ["Object"] = "Object", ["Array"] = "Array", ["Function"] = "Function" }

array_reduce = function (callback_function, an_array, initial_value)
    -- JavaScript-like Array.reduce() function
    local result = initial_value
    for array_item_index, array_item in ipairs(an_array) do
        result = callback_function(result, array_item, array_item_index, an_array)
    end
    return result
end

string_repeat = function (a_string, times)
    local result = ""
    for i = 1, times, 1 do -- start, stop, step
        result = result .. a_string
    end
    return result
end

ternary = function (true_condition, callback_function_if_condition_true, callback_function_if_condition_false)
    if (true_condition == true) then return callback_function_if_condition_true() end
    return callback_function_if_condition_false()
end

is_like_js_null = function (anything) return (type(anything) == "nil") end

is_like_js_boolean = function (anything) return (type(anything) == "boolean") end

is_like_js_string = function (anything) return (type(anything) == "string") end

is_like_js_numeric = function (anything) return (type(anything) == "number") end

is_like_js_function = function (anything) return (type(anything) == "function") end

is_like_js_object = function (anything)
    if (type(anything) ~= "table") then return false end
    for object_key, object_value in pairs(anything) do
        if ((type(object_key) == "number") and ((object_key >= 1) and (object_key <= #anything))) then return false end
    end
    if (next(anything) == nil) then return false end
    return true
end

is_like_js_array = function (anything)
    if (type(anything) ~= "table") then return false end
    for object_key, object_value in pairs(anything) do
        if ((type(object_key) == "number") and ((object_key >= 1) and (object_key <= #anything))) then return true end
    end
    if (next(anything) == nil) then return true end
    return false
end

get_type = function (anything) return (
    ternary((is_like_js_null(anything) == true),
        (function () return js_like_type["Null"] end),
        (function () return (
            ternary((is_like_js_boolean(anything) == true),
                (function () return js_like_type["Boolean"] end),
                (function () return (
                    ternary((is_like_js_string(anything) == true),
                        (function () return js_like_type["String"] end),
                        (function () return (
                            ternary((is_like_js_numeric(anything) == true),
                                (function () return js_like_type["Numeric"] end),
                                (function () return (
                                    ternary((is_like_js_object(anything) == true),
                                        (function () return js_like_type["Object"] end),
                                        (function () return (
                                            ternary((is_like_js_array(anything) == true),
                                                (function () return js_like_type["Array"] end),
                                                (function () return (
                                                    ternary((is_like_js_function(anything) == true),
                                                        (function () return js_like_type["Function"] end),
                                                        (function () return type(anything) end)
                                                    )
                                                ) end)
                                            )
                                        ) end)
                                    )
                                ) end)
                            )
                        ) end)
                    )
                ) end)
            )
        ) end)
    )
) end

optional_chaining = function (callback_function)
    local is_not_error, result = xpcall(callback_function, (function (an_error_message) return nil end))
    return result
end

nullish_coalescing = function (anything, default_value) return (ternary((get_type(anything) == js_like_type["Null"]), (function () return default_value end), (function () return anything end))) end

object_keys = function (an_object)
    local new_array = {}
    for object_key, object_value in pairs(an_object) do
        table.insert(new_array, object_key)
    end
    return new_array
end

json_stringify = function (anything, optional_keyword_argument_object)
    optional_keyword_argument_object = nullish_coalescing(optional_keyword_argument_object, {})
    local pretty = optional_chaining(function () return optional_keyword_argument_object["pretty"] end)
    pretty = ternary((get_type(pretty) == js_like_type["Null"]), (function () return false end), (function () return pretty end))
    local indent_default = string_repeat(" ", 4)
    local indent_level = 0
    local json_stringify_inner
    json_stringify_inner = function (anything_inner)
        local anything_inner_type = get_type(anything_inner)
        if (anything_inner_type == js_like_type["Null"]) then return "null" end
        if (anything_inner_type == js_like_type["String"]) then return ('"' .. anything_inner .. '"') end
        if ((anything_inner_type == js_like_type["Numeric"]) or (anything_inner_type == js_like_type["Boolean"])) then return tostring(anything_inner) end
        if (anything_inner_type == js_like_type["Object"]) then
            local object_keys_array_length = #object_keys(anything_inner)
            if (object_keys_array_length == 0) then return "{}" end
            indent_level = indent_level + 1
            local result = ternary((pretty == true), (function () return ("{\n" .. string_repeat(indent_default, indent_level)) end), (function () return "{ " end))
            local object_entry_index = 1
            for object_key, object_value in pairs(anything_inner) do
                result = result .. '"' .. object_key .. '": ' .. json_stringify_inner(object_value)
                if (object_entry_index ~= object_keys_array_length) then
                    result = result .. ternary((pretty == true), (function () return (",\n" .. string_repeat(indent_default, indent_level)) end), (function () return ", " end))
                end
                object_entry_index = object_entry_index + 1
            end
            indent_level = indent_level - 1
            result = result .. ternary((pretty == true), (function () return ("\n" .. string_repeat(indent_default, indent_level) .. "}") end), (function () return " }" end))
            return result
        end
        if (anything_inner_type == js_like_type["Array"]) then
            if (#anything_inner == 0) then return "[]" end
            indent_level = indent_level + 1
            local result = ternary((pretty == true), (function () return ("[\n" .. string_repeat(indent_default, indent_level)) end), (function () return "[" end))
            for array_item_index, array_item in ipairs(anything_inner) do
                result = result .. json_stringify_inner(array_item)
                if (array_item_index ~= #anything_inner) then
                    result = result .. ternary((pretty == true), (function () return (",\n" .. string_repeat(indent_default, indent_level)) end), (function () return ", " end))
                end
            end
            indent_level = indent_level - 1
            result = result .. ternary((pretty == true), (function () return ("\n" .. string_repeat(indent_default, indent_level) .. "]") end), (function () return "]" end))
            return result
        end
        if (anything_inner_type == js_like_type["Function"]) then return "[object Function]" end
        return anything_inner_type
    end
    return json_stringify_inner(anything)
end

string_interpolation = function (...)
    local rest_arguments = {...}
    return (array_reduce(function (current_result, current_argument)
        local current_argument_type = get_type(current_argument)
        return (current_result .. (ternary((current_argument_type == js_like_type["String"]), (function () return current_argument end), (function () return (ternary(((current_argument_type == js_like_type["Array"]) and (#current_argument == 1)), (function () return json_stringify(optional_chaining(function () return current_argument[1] end)) end), (function () return json_stringify(current_argument) end))) end))))
    end, rest_arguments, ""))
end

console_log = function (...)
    print(string_interpolation(...))
end

something = "foo"
console_log(string_interpolation("something: ", json_stringify(something, { ["pretty"] = true })))
something = 123
console_log(string_interpolation("something: ", json_stringify(something, { ["pretty"] = true })))
something = true
console_log(string_interpolation("something: ", json_stringify(something, { ["pretty"] = true })))
something = nil
console_log(string_interpolation("something: ", json_stringify(something, { ["pretty"] = true })))
something = {1, 2, 3}
console_log(string_interpolation("something: ", json_stringify(something, { ["pretty"] = true })))
something = { ["foo"] = "bar" }
console_log(string_interpolation("something: ", json_stringify(something, { ["pretty"] = true })))
